Add `populate_backlog` rpc command

pwojcikdev opened the pull request:
This PR adds a new rpc command for populating backlog. Populating backlog is a process in the node that scans all accounts , checks for unconfirmed blocks in that account's chain and queues those blocks for confirmation via election scheduler. I found this command to be very useful when setting up local test networks, as by default backlog population is done in 5 minute intervals which is too slow. It is no doubt that a better way for scheduling the backlog population would be a better long term solution, but I think a manual way to trigger that is nevertheless very useful.

Request:
`{
  "action": "populate_backlog"  
}`

Response:
`{
  "success": ""
}`

pwojcikdev committed:
Add `populate_backlog` rpc

pwojcikdev committed:
Add `populate_backlog` rpc tests

dsiganos commented on nano/node/json_handler.cpp:
```diff
@@ -5224,6 +5224,13 @@ void nano::json_handler::work_peers_clear ()
 
+void nano::json_handler::populate_backlog ()
+{
+	node.populate_backlog ();
```
I think rather than calling populate backlog directly, it should just trigger one.
I do not think we should be doing heavy processing in the RPC handler.

pwojcikdev replied:
The problem with triggering it to be run in the background is that way it would be possible to block all `node.workers` and node needs those for realtime processing. This way it is only the rpc that will stop responding for a moment, not the whole node.

pwojcikdev replied:
We could move the backlog population to a dedicated thread and trigger it there (in a dedicated class perhaps), that sounds reasonable to me.

dsiganos replied:
That sounds reasonable to me too.

pwojcikdev committed:
Only allow `populate_backlog` RPC under enable_control

pwojcikdev committed:
Extract backlog population logic to separate class and run it on a dedicated thread

pwojcikdev committed:
Delay backlog population initial start

dsiganos commented on nano/node/backlog_population.cpp:
```diff
@@ -0,0 +1,90 @@
+nano::backlog_population::~backlog_population ()
+{
+	stop ();
+	if (thread.joinable ())
```
Is this thread safe?

dsiganos replied:
Ignoring this as we use this pattern in other places.

dsiganos commented on nano/node/backlog_population.cpp:
```diff
@@ -0,0 +1,90 @@
+			lock.unlock ();
+			populate_backlog ();
+			lock.lock ();
+			triggered = false;
```
This should be moved to before calling  populate_backlog() so that a trigger cannot possibly be missed in the case that trigger comes in at the end of the populate backlog cycle.

dsiganos commented on nano/node/backlog_population.cpp:
```diff
@@ -0,0 +1,90 @@
+	nano::unique_lock<nano::mutex> lock{ mutex };
+	while (!stopped)
+	{
+		if (predicate () || (node.config.frontiers_confirmation != nano::frontiers_confirmation_mode::disabled))
```
Why is this check 
```node.config.frontiers_confirmation != nano::frontiers_confirmation_mode::disabled```
not inside `predicate()`? I think it should be either inside predicate or the check should be outside the loop. Currently, the loop checks for this at every loop pass but it won't consider it for wakeup from sleep, which seems inconsistent.

dsiganos replied:
I thought that the frontiers_confirmation_mode was being read inside the loop so that the backlog_population object could auto-adjust to live changes to the config value `node.config.frontiers_confirmation`. but it appears that the loop is always running, whether frontiers confirmation is enabled or not, and the loop checks if it was manually triggered or it is ongoing mode. So this isn't an issue, it was a misunderstanding.

dsiganos commented on nano/node/backlog_population.cpp:
```diff
@@ -0,0 +1,90 @@
+	{
+		auto transaction = node.store.tx_begin_read ();
+		auto count = 0;
+		for (auto i = node.store.account.begin (transaction, next), n = node.store.account.end (); !stopped && i != n && count < chunk_size; ++i, ++count, ++total)
```
This line is not easily readable for no good reason.
The initializations could be moved outside the for loop and dramatically reduce the width of that line.

pwojcikdev replied:
Moved as-is from the previous code, doing a quick scan there are multiple places where we use this long line pattern. This has also the benefit of keeping the variable in a tighter scope.

clemahieu replied:
There are a lot of wide lines in the codebase I don't think this should hold it up.

dsiganos replied:
I'd argue that new code should be held to higher standards than existing code but I agree that we do not need to hold up the Pr for a minor aesthetic change.

pwojcikdev replied:
Just enable world wrapping in your editor, I personally preferred the old line it seems cleaner for me.

pwojcikdev replied:
If we want to enforce line width limits that should go to our clang format config file, in a separate PR

dsiganos committed:
Move delay assignment outside the loop and make it constant.

dsiganos committed:
Document backlog_population::triggered and thread members

It is not immediately obvious that the trigger can be used even when
backlog population is disabled and that it is used for manual triggering
mainly.

dsiganos committed:
Improve code readability

pwojcikdev committed:
Move triggered to start of the loop

pwojcikdev committed:
Use explicit dependencies

dsiganos committed:
Rename store to store_m to avoid compilation problem on gcc

dsiganos committed:
Document intention of notify function

dsiganos committed:
Break backlog_population's dependency on nodeconfig

Introduce the class nano::backlog_population::config to hold the
configuration items and not need access to nano::nodeconfig.

dsiganos committed:
Fix for compilation problem.

pwojcikdev committed:
Move to namespace
